### PR TITLE
no-color cli output

### PR DESCRIPTION
**Pull request by dharmateja03**
## Type of Change

- [ ] Bug fix
- [ X ] New feature
- [ ] Refactor
- [ ] Documentation

## Checklist
- [x] npm run lint passes
  - [x] npx tsc --noEmit passes
  - [x] npm run test:run passes
  - [x] npm run build succeeds

## Release Notes

* **New Features**
  * Added `NO_COLOR` environment variable and `--no-color` command-line flag to disable ANSI-colored CLI output for improved compatibility with CI logs and non-ANSI-compatible terminals.

* **Documentation**
  * Updated documentation to describe the new color-disabling options.

### CODE REVIEW

**Comment by NiveditJain** [comment]
Thanks for the PR @dharmateja03! Curious about the motivation here — are you running failproofai somewhere with NO_COLOR set (CI logs, a piped tool, a non-ANSI terminal), or was this more of a general spec-compliance improvement? Just want to understand the use case so we can make sure the no-color path covers what you actually need.

File: __tests__/hooks/install-prompt.test.ts [hunk 1]
```diff
@@ -131,5 +158,32 @@ describe("hooks/install-prompt", () => {
...
+      vi.resetModules();
+      const { resolveTargetClis } = await import("../../src/hooks/install-prompt");
+      await resolveTargetClis(undefined, "install");
+      spy.mockRestore();
+      vi.doUnmock("../../src/hooks/integrations");
+      vi.resetModules();
+      expect(logs[0]).toMatchInlineSnapshot(
+        "\"Warning: no agent CLI binary found in PATH (claude, codex, copilot, cursor-agent). Defaulting to Claude Code; hooks will activate when an agent is installed.\"",
+      );
+      expect(logs[0]).not.toContain("\\u001b[");
```
Comment (NiveditJain) [inline __tests__/hooks/install-prompt.test.ts:186]: 🟡 **Warning — this assertion can never fail, even for fully-colored output.**

In JS source, `"\\u001b["` is the **7 literal characters** `\ u 0 0 1 b [`, not the actual ANSI escape byte `0x1B`. A real ANSI string contains the escape *byte* (`ESC[33m…`), which never contains that literal 7-char substring. So this check passes even when `logs[0]` is fully colored — it is a no-op.

Proven locally:

```
real string contains ESC byte:                          true
real string contains literal backslash-u001b :          false
=> .not.toContain("\\u001b[") passes on a FULLY-COLORED string
```

The `toMatchInlineSnapshot` on line 183 is what actually guards against ANSI here; this line adds only false confidence.

**Fix:** use a single backslash so it matches the real escape char:

```ts
expect(logs[0]).not.toContain("[");   // or "\x1b["
```

File: src/cli-color.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  let hidden = false;
+  const show = (): void => {
+    if (!enabled || !hidden) return;
+    stream.write("\x1B[?25h");
+    hidden = false;
+  };
+  const onExit = (): void => {
+    show();
+  };
+  process.on("exit", onExit);
```
Comment (NiveditJain) [inline src/cli-color.ts:73]: 🔵 **Info — `process.on("exit")` is registered unconditionally, even when the controller is disabled.**

`enabled` is computed on line 63, but the exit listener is added on every call regardless of it. When `enabled` is `false` (non-TTY, `NO_COLOR`, `--no-color`) the controller's `hide`/`show` are no-ops, yet a `process.on("exit")` listener is still attached and only removed if the caller reaches `dispose()`.

In this PR's flow that's harmless (controllers are only created in the interactive path after the `!isTTY || !style.enabled` early-returns, and `dispose()` runs in `cleanup()`). But this is a shared utility — a future caller that creates a controller in the disabled state without disposing would leak a listener, and repeated creation can trip Node's `MaxListenersExceededWarning`.

**Fix:** only attach the exit handler when it can do something:

```ts
if (enabled) process.on("exit", onExit);
```

and guard the `removeListener` in `dispose()` accordingly.

**Comment by NiveditJain** [comment]
## 🔍 Hermes Code Review

### 📋 Executive Summary
This PR adds `NO_COLOR` / `--no-color` support: a new `src/cli-color.ts` module centralizes ANSI styling + cursor control, and `src/hooks/install-prompt.ts` is refactored to use it (falling back to non-interactive defaults when color is disabled). The refactor itself is clean, well-typed (`tsc --noEmit` passes), and its 8 unit tests pass. **However, the feature is incomplete**: the most prominent colored output — the `policies` status table and warnings in `src/hooks/manager.ts` — is not gated, so ANSI codes still leak with `NO_COLOR=1`. The branch is also **`CONFLICTING` with `main`** and cannot be merged as-is.

---

### 📊 Change Architecture

```mermaid
graph TD
    A["bin/failproofai.mjs<br/>strips --no-color → NO_COLOR=1"] -->|sets env| B["src/cli-color.ts (NEW)<br/>isColorEnabled / createCliStyler / createCursorController"]
    B -->|"used by"| C["src/hooks/install-prompt.ts<br/>promptPolicySelection / resolveTargetClis"]
    C -->|"style.enabled === false"| D["non-interactive defaults ✅"]
    B -.->|"NOT wired in ❌"| E["src/hooks/manager.ts<br/>policies list + warnings<br/>13 hardcoded ANSI sites"]
    E -->|"NO_COLOR=1 → 32 ESC bytes still emitted"| F["Terminal output still colored 🔴"]
    style B fill:#90EE90
    style C fill:#87CEEB
    style E fill:#FFD700
    style F fill:#FFB6C1
```

**Legend:** 🟢 New | 🔵 Modified | 🟡 Gap (not wired up) | 🔴 Observed defect

---

### 🔴 Breaking Changes
✅ No API/schema/config breaking changes. New env var + flag are purely additive.

⚠️ **Merge blocker (process):** the branch is `mergeable: CONFLICTING` / `mergeStateStatus: DIRTY`. It is dozens of commits behind `origin/main` (multiple releases incl. `0.0.11` and `0.0.12-beta.0`, and many dependency bumps). `CHANGELOG.md` and `package.json` will conflict. Per this repo's `CLAUDE.md` ("Branch must contain all commits from main"), rebase onto `origin/main` before merge. Note the green CI ran on 2026-04-30 against a now-stale base.

---

### ⚠️ Issues Found
1. 🟡 **Major (incomplete feature)** — `src/hooks/manager.ts` (13 sites: L256, 468, 474, 533, 546, 555, 563, 572, 576, 580, 604, 608, 612) — the `policies` list table and warnings still emit raw ANSI under `NO_COLOR`. **Not routed through `cli-color`.** The CHANGELOG claim "ANSI styling is suppressed for warnings" is only partially true.
2. 🟡 **Warning (test quality)** — `__tests__/hooks/install-prompt.test.ts:186` — `expect(logs[0]).not.toContain("\\u001b[")` matches a literal 7-char string, not the ESC byte; it passes even for fully-colored output (no-op assertion). See inline comment.
3. 🔵 **Info (robustness)** — `src/cli-color.ts:73` — `process.on("exit")` is attached even when the controller is disabled; harmless in this PR's flow but a latent listener leak for future callers. See inline comment.
4. 🔵 **Info (process)** — branch name `fix/ci-stop-current-head` is unrelated to the feature ("no-color cli output"), and the PR description body is an empty template. Minor traceability nit (also flagged by CodeRabbit).

---

### 🔬 Logical / Bug Analysis
- **`isColorEnabled` precedence is correct.** `NO_COLOR` (any non-empty value) → off; empty string does *not* disable (spec-compliant); `FORCE_COLOR` truthy → on; else require TTY. `--no-color` + `FORCE_COLOR` together correctly resolve to *off* because the bin sets `NO_COLOR=1` and that check runs first.
- **`hasNoColorFlag()` reads `process.argv`** — `bin` filters `--no-color` out of its local `args` but never mutates `process.argv`, so the flag remains detectable. Redundant with the env var but harmless.
- **`promptPolicySelection` / `resolveTargetClis` fallbacks are sound** — `!process.stdin.isTTY || !style.enabled` correctly returns defaults / all-detected. Controllers are created only after these early returns, so no raw-mode/cursor manipulation happens in no-color mode (verified by the added tests).
- No race conditions, off-by-one, null-deref, or resource-leak issues found in the diffed code paths.

---

### 🧪 Evidence — Build & Test Results

<details>
<summary>Typecheck + unit tests (this branch)</summary>

```
$ npx tsc --noEmit        → exit 0

$ bun run test:run __tests__/hooks/install-prompt.test.ts
 Test Files  1 passed (1)
      Tests  8 passed (8)
```

</details>

<details>
<summary>🔴 Incomplete NO_COLOR coverage — reproduced end-to-end</summary>

Built `dist/index.js` and ran the `policies` list view with color disabled, counting ANSI ESC (0x1B) bytes in the output:

```
$ NO_COLOR=1  failproofai policies   →  32 ESC bytes still emitted
```

The status column checkmarks render as `ESC[32m✓ESC[0m` and the separators as `ESC[2m──ESC[0m` despite `NO_COLOR=1`. Source: `src/hooks/manager.ts` contains **13** hardcoded `\x1B[...` sites that this PR does not route through `cli-color`.
```
$ grep -c '\\x1B\[' src/hooks/manager.ts   →  13
```

</details>

<details>
<summary>Weak-assertion proof (test line 186)</summary>

```
real string contains ESC byte:                       true
real string contains literal backslash-u001b[ :      false
=> .not.toContain("\\u001b[") passes on a FULLY-COLORED string
```

</details>

---

### 🔗 Issue Linkage
No issue is linked via GitHub's "closes" mechanism. CodeRabbit associates this with the upstream `NO_COLOR` request (#251). The maintainer's open question about the concrete use case (CI logs vs. spec-compliance) is still unanswered — worth resolving, since it determines whether the `manager.ts` output must also be covered (for CI logs, it does).

---

### 💡 Suggestions
- **Finish the feature:** thread `createCliStyler(process.stdout)` through `src/hooks/manager.ts` so the `policies` table, beta separators, and warnings honor `NO_COLOR`. This is the output users most often pipe into CI logs — the stated motivation for the flag. Add a regression test asserting zero ESC bytes in the list output under `NO_COLOR`.
- Consider a single shared module-level styler (or a `stripAnsi` helper) rather than re-deriving `createCliStyler` per function, to prevent this kind of drift.
- Fill in the PR description and align the branch name with the change for traceability.

---

### 🏆 Verdict

🔴 **CHANGES REQUESTED**

- **Blocker:** branch is `CONFLICTING` with `main` — rebase required before merge.
- **Major:** the advertised NO_COLOR behavior is incomplete — `policies` list + warnings (`manager.ts`, 13 sites) still emit ANSI (reproduced: 32 ESC bytes under `NO_COLOR=1`).
- **Minor:** ineffective test assertion (test:186) and an unconditional exit-listener (cli-color.ts:73), both noted inline.

The refactor is a good, clean foundation — it just needs to be applied to the rest of the CLI output before the feature actually delivers on its title.

---

*Reviewed by Hermes · Claude Opus 4.8 · 2026-07-04*

**Review by NiveditJain — CHANGES_REQUESTED** [review]
Hermes review — 🔴 **Changes requested.** Three items, details in the summary comment and inline notes:

1. **Merge blocker:** branch is `CONFLICTING` with `main` (behind by multiple releases + dep bumps) — rebase onto `origin/main` first.
2. **Major (incomplete feature):** `policies` list + warnings in `src/hooks/manager.ts` (13 hardcoded ANSI sites) are not routed through the new `cli-color` module, so ANSI still leaks under `NO_COLOR` — reproduced end-to-end (32 ESC bytes emitted with `NO_COLOR=1`).
3. **Minor:** the `not.toContain("\\u001b[")` assertion (test:186) is a no-op, and `createCursorController` attaches a `process.on("exit")` listener even when disabled (cli-color.ts:73).

The `install-prompt.ts` refactor itself is clean and well-tested — nice foundation; it just needs to cover the rest of the CLI output.

**Comment by chhhee10** [comment]
Thanks for this, and sorry it sat so long. Closing it — the branch is 350 commits behind, and `main` has since grown `src/hooks/tui.ts`, a shared painter that already gates on `NO_COLOR`, so `cli-color.ts` would duplicate it.

You were right about the bug, though — `manager.ts` still leaks 16 hardcoded ANSI sites under `NO_COLOR=1`.

Much smaller job now if you ever want another go: route those through `paint()` and add `--no-color` in `bin/failproofai.mjs`. Happy to review it.
